### PR TITLE
Adding CI/CD + GitOps draft

### PR DESCRIPTION
**Pull request by kalexand-rh**

### CODE REVIEW

**Comment by kalexand-rh** [comment]
@siamaksade, @sferich888, this is a draft of general CI/CD and GitOps architecture content that Ali's been working on. Will you PTAL?

**Comment by kalexand-rh** [comment]
@openshift/team-documentation, PTAL?

File: architecture/cicd_gitops.adoc [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+_Continuous integration_ is an automation process for developers. Code changes to an application are regularly built, tested, and merged to a shared repository.
+
+_Continuous delivery_ and _continuous deployment_ are closely related concepts that are sometimes used interchangeably that refer to automation of the pipeline.
+Continuous delivery uses automation to ensure that a developer’s changes to an application are tested and sent to a repository, where an operation team can deploy them to a production environment. Continuous deployment enables the release of changes, starting from the repository until production. Continuous deployment prevents the operations team from getting overloaded and speeds up application delivery.
+
+[id="cicd_gitops_methodology"]
+== The GitOps methodology and practice
+
+_GitOps_ is a set of practices that use Git pull requests to manage infrastructure and application configurations. The Git repository in GitOps is the only source of truth and contains the entire state of the system so that the trail of changes to the system state are visible and auditable. GitOps enables you to implement a DevOps methodology.
```
Comment (bobfuru) [inline architecture/cicd_gitops.adoc:23]: "only source of truth" for what?

File: architecture/cicd_gitops.adoc [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+You can use GitOps tooling to create repeatable and predictable processes for managing and recreating {product-title} clusters and applications. By using GitOps, you can address the issues of infrastructure and application configurations sprawl. It simplifies the propagation of infrastructure and application configurations changes across multiple clusters by defining your infrastructure and applications definitions as “code”. Implementing GitOps for your cluster configuration files can make automated installation easier and allow you to configure automated cluster customizations. You can apply the core principles of developing and maintaining software in a Git repo to the creation and management of your cluster and application configuration files.
```
Comment (bobfuru) [inline architecture/cicd_gitops.adoc:25]: s/configurations sprawl/configuration sprawl/
s/application configurations changes/application configuration changes/

File: architecture/cicd_gitops.adoc [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+Automating both your cluster configuration and container development process by using GitOps allows you to pick and choose where and when to adopt this practice, using {product-title}. Using a CI pipeline that pairs with your GitOps strategy and execution plan is ideal. {product-title} provides the flexibility to choose when and how you integrate this methodology into your business practices and pipelines.
```
Comment (bobfuru) [inline architecture/cicd_gitops.adoc:27]: "adopt this practice" - What practice? The GitOps practice?
"using OCP" sounds oddly attached here. Is your point that OCP integrates GitOps or that GitOps makes OCP easier? Not really sure.

"Using a CI pipeline that pairs with your GitOps strategy and execution plan is ideal." Why? I don't think this sentence strengthens your argument, which (IIUC) is that OCP's flexibility lets you integrate CI/CD for better business practices.

File: architecture/cicd_gitops.adoc [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+GitOps works well with {product-title} because you can declaratively configure it and the state of the cluster configuration can be stored using Git. For more information, see xref:../installing/install_config/customizations.adoc#customizations[Available cluster customizations].
```
Comment (bobfuru) [inline architecture/cicd_gitops.adoc:29]: "configure it" - what, GitOps or OCP?
Suggest a stronger opening sentence here than "works well" - maybe something like:
With GitOps integration, you can declaratively configure and store your OCP cluster configurations.

File: architecture/cicd_gitops.adoc [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+[id="cicd_gitops_cluster_administration"]
+=== GitOps for single-cluster and multi-cluster administration
+
+Whether you need one or multiple independent or co-operative {product-title} clusters, you can use a GitOps strategy to manage the following tasks:
```
Comment (bobfuru) [inline architecture/cicd_gitops.adoc:34]: s/Whether you need one or multiple independent or co-operative {product-title} clusters,/Whether you need one or more independent or cooperative clusters,/

Comment (bobfuru) [inline architecture/cicd_gitops.adoc:34]: s/you can use a GitOps strategy to manage the following tasks:/you can use a GitOps strategy to:

File: architecture/cicd_gitops.adoc [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+* Ensure that these clusters have similar states for configuration, monitoring, or storage.
```
Comment (bobfuru) [inline architecture/cicd_gitops.adoc:36]: s/these/your/ (or "the")

File: architecture/cicd_gitops.adoc [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+* Recover or recreate clusters from a known state.
```
Comment (bobfuru) [inline architecture/cicd_gitops.adoc:37]: "Recreate or recover" sounds better in my head (though my head should seldom be trusted lol).

Comment (bobfuru) [reply architecture/cicd_gitops.adoc:37]: Actually, I'd combine this bullet and the next one: "Create, recreate, or recover clusters..." (no need for two bullets imo).

File: architecture/cicd_gitops.adoc [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+* Create clusters with a known state.
+* Apply configuration changes to multiple {product-title} clusters.
+* Revert configuration changes to multiple {product-title} clusters.
```
Comment (bobfuru) [inline architecture/cicd_gitops.adoc:40]: Same as above comment. I'd combine these two: "Apply or revert..."

File: architecture/cicd_gitops.adoc [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+* Associate templated configuration with different environments.
```
Comment (bobfuru) [inline architecture/cicd_gitops.adoc:41]: In other words, "Match configuration templates to different environments"?

File: architecture/cicd_gitops.adoc [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+
+[id="cicd_gitops_application_configuration"]
+=== GitOps for application configuration management
+
+You can also use GitOps practices to manage application configuration. This practice ensures consistency in applications when you must deploy them to different clusters in different environments, like development, stage, and production, during their life cycle. Managing application configuration with GitOps is also beneficial when you must deploy applications across multiple clusters, whether on-cloud or on-premise, for availability and scalability purposes.
```
Comment (bobfuru) [inline architecture/cicd_gitops.adoc:46]: I'd shorten to something like, "This ensures consistency when deploying applications in (on?) clusters in different environments, such as development, stage, and production.

Comment (bobfuru) [inline architecture/cicd_gitops.adoc:46]: What's the latest ruling on our usage of on-premise? I think this is it? I haven't seen "on-cloud" before. Maybe this sentence could be shortened to something like:
GitOps is also beneficial when deploying multiple cloud or on-premise clusters for greater availability and scalability.

File: architecture/cicd_gitops.adoc [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+
+You can use a GitOps strategy to:
```
Comment (bobfuru) [inline architecture/cicd_gitops.adoc:48]: I don't see this list as adding anything above what was highlighted in the previous paragraph. I suggest removing it.

File: architecture/cicd_gitops.adoc [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+* Promote applications across clusters, from stage to production.
+* Apply application changes to multiple OCP clusters.
+* Revert application changes to multiple OCP clusters.
+
+[id="cicd_gitops_integrators"]
+=== GitOps technology providers and integrators
+
+There are several community offerings and third-party vendors that provide a high level of integration with {product-title}.
+
+You can integrate GitOps into {product-title} with the following community partners and third party integrators:
```
Comment (bobfuru) [inline architecture/cicd_gitops.adoc:59]: s/third party integrators/third party-integrators/

File: architecture/cicd_gitops.adoc [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+
+* ArgoCD
```
Comment (bobfuru) [inline architecture/cicd_gitops.adoc:61]: I see this setup as useful if we plan to introduce additional integrators. But if not, I think it's overboard to have this in a list. Suggest a single sentence:
There are several community offerings and third-party vendors that provide a high level of integration with OpenShift Container Platform, such as ArgoCD.

**Review by bobfuru — COMMENTED** [review]
Overall, this is a great article with good high-level and conceptual content. Most of my comments are suggestions for tightening up the message and cutting out a few redundancies. Nice work! If you want me to take another look, let me know. Otherwise, once you've applied the changes that make sense to you, this LGTM.

**Pull request merged**